### PR TITLE
[SPARK-47447][SQL] Allow reading Parquet TimestampLTZ as TimestampNTZ

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Currently, Parquet TimestampNTZ type columns can be read as TimestampLTZ, while reading TimestampLTZ as TimestampNTZ will cause errors. This makes it impossible to read parquet files containing both TimestampLTZ and TimestampNTZ as TimestampNTZ.

To make the data type system on Parquet simpler, this PR allows reading TimestampLTZ as TimestampNTZ in the Parquet data source.

### Why are the changes needed?

* Make it possible  to read parquet files containing both TimestampLTZ and TimestampNTZ as TimestampNTZ
* Make the data type system on Parquet simpler

### Does this PR introduce _any_ user-facing change?

Yes, Parquet TimestampLTZ type column are now allowed to be read as TimestampNTZ

### How was this patch tested?

UT
### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @sadikovi @srielau too

**Comment by yaooqinn** [comment]
Could you explain why we make a difference in the way Parquet and JDBC handle LTZ? As for JDBC LTZ, this kind of conversion is forbidden if we know we are reading an LTZ value

**Comment by cloud-fan** [comment]
due to schema evolution, it's very likely to have mixed data types in parquet files. e.g. turn on NTZ, write some files, turn it off (maybe hit some issues), write some files, and turn it on again (issue fixed), read some files.

**Comment by gengliangwang** [comment]
@yaooqinn For parquet files, it's possible that a file directory contains two different kinds of timestamp columns. Before this change, only timestamp type can be used to read such a file directory. While for JDBC, a table type is certain. 
But you make a good point here. I am also considering allowing reading INT96 as TimestampNTZ, since there is no clear definition of this Parquet physical type. We can still fobid reading Int64 timestamps with `isAdjustedToUTC=true` as TimestampNTZ.

**Comment by yaooqinn** [comment]
"Understood. Thank you for explaining that, @gengliangwang."

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala [hunk 1]
```diff
@@ -436,6 +436,17 @@ private[parquet] class ParquetRowConverter(
           }
         }
 
+      // INT96 timestamp doesn't have a logical type, here we check the physical type instead.
+      case TimestampNTZType if parquetType.asPrimitiveType().getPrimitiveTypeName == INT96 =>
+        new ParquetPrimitiveConverter(updater) {
+          // Converts nanosecond timestamps stored as INT96.
+          // TimestampNTZ type does not require rebasing due to its lack of time zone context.
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala:443]: Could you tell me why we can treat ltz as ntz directly?

Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala:443]: * LTZ doesn't really store the time zone info in Parquet files.  Also, Spark uses the long value directly when reading NTZ as LTZ. I am trying to make it simple and Symmetrical.
* If we shift by sessional time zone here, probably we need to do it when reading NTZ as LTZ, which will be a breaking change. Also, the result of NTZ columns will be affected by the sessional time zone conf `spark.sql.session.timeZone`.

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala:443]: Got it. Thank you!

**Comment by gengliangwang** [comment]
@yaooqinn @cloud-fan @beliefer Thanks for the review. I am merging this one to master.

**Pull request closed without merging**